### PR TITLE
[config] support for configuring muxcable to manual mode of operation 

### PR DESCRIPTION
**Pull request by vdahiya12**
Signed-off-by: vaibhav-dahiya <vdahiya@microsoft.com>

#### What I did
This PR adds support for an option to  configure muxcable mode to  a `manual` mode. The manual mode is in addition
to `auto\active` mode. 

The new output would look like this in case an active flag is passed to the command line

`admin@sonic:~$ sudo config muxcable mode manual Ethernet0`

`admin@sonic:~$ sudo config muxcable mode manual all`

added an option to set muxcable  mode to manual mode, in addition to existing auto/active modes. 

#### How I did it

added the changes in config/muxcable.py and added testcases

#### How to verify it
Ran the unit tests

### CODE REVIEW

File: config/muxcable.py [hunk 1]
```diff
@@ -73,43 +73,48 @@ def lookup_statedb_and_update_configdb(per_npu_statedb, config_db, port, state_c
...
     state = get_value_for_key_in_dict(muxcable_statedb_dict, port, "state", "MUX_CABLE_TABLE")
-    if (state == "active" and configdb_state == "active") or (state == "standby" and configdb_state == "active") or (state == "unknown" and configdb_state == "active"):
-        if state_cfg_val == "active":
-            # status is already active, so right back error
+    if (state == "active" and configdb_state == "active") or (state == "standby" and configdb_state == "active") or (state == "unknown" and configdb_state == "active") or (state == "active" and configdb_state == "manual") or (state == "standby" and configdb_state == "manual") or (state == "unknown" and configdb_state == "manual"):
+        if state_cfg_val == configdb_state:
+            # status is already active or manual, so nothing to do
             port_status_dict[port] = 'OK'
-        if state_cfg_val == "auto":
+        elif state_cfg_val == "auto" or state_cfg_val == "manual":
```
Comment (jleveque) [inline config/muxcable.py:80]: Suggest simplifying:
```python
elif state_cfg_val in ["auto", "manual"]:
```

File: config/muxcable.py [hunk 2]
```diff
@@ -73,43 +73,48 @@ def lookup_statedb_and_update_configdb(per_npu_statedb, config_db, port, state_c
...
             # display ok and write to cfgdb auto
             port_status_dict[port] = 'OK'
-            config_db.set_entry("MUX_CABLE", port, {"state": "auto",
+            config_db.set_entry("MUX_CABLE", port, {"state": state_cfg_val,
                                                     "server_ipv4": ipv4_value, "server_ipv6": ipv6_value})
-    elif state == "active" and configdb_state == "auto":
-        if state_cfg_val == "active":
+    elif state == "active" and configdb_state == "auto" or state == "active" and configdb_state == "manual":
```
Comment (jleveque) [inline config/muxcable.py:85]: Suggest simplifying:
```python
elif state == "active" and configdb_state in ["auto", "manual"]:
```

File: config/muxcable.py [hunk 3]
```diff
@@ -73,43 +73,48 @@ def lookup_statedb_and_update_configdb(per_npu_statedb, config_db, port, state_c
...
+        if state_cfg_val == configdb_state:
+            # dont write anything to db, write OK to user
+            port_status_dict[port] = 'OK'
+        elif state_cfg_val == "active" or state_cfg_val == "manual":
```
Comment (jleveque) [inline config/muxcable.py:89]: Suggest simplifying:
```python
elif state_cfg_val in ["auto", "manual"]:
```

File: config/muxcable.py [hunk 4]
```diff
@@ -73,43 +73,48 @@ def lookup_statedb_and_update_configdb(per_npu_statedb, config_db, port, state_c
...
         if state_cfg_val == "active":
             # make the state active
             config_db.set_entry("MUX_CABLE", port, {"state": "active",
                                                     "server_ipv4": ipv4_value, "server_ipv6": ipv6_value})
             port_status_dict[port] = 'INPROGRESS'
-        if state_cfg_val == "auto":
+        elif state_cfg_val == configdb_state:
             # dont write anything to db
             port_status_dict[port] = 'OK'
+        elif state_cfg_val == "manual" or state_cfg_val == "auto":
```
Comment (jleveque) [inline config/muxcable.py:104]: Suggest simplifying:
```python
elif state_cfg_val in ["auto", "manual"]:
```

File: config/muxcable.py [hunk 5]
```diff
@@ -73,43 +73,48 @@ def lookup_statedb_and_update_configdb(per_npu_statedb, config_db, port, state_c
...
-            config_db.set_entry("MUX_CABLE", port, {"state": "active",
+            config_db.set_entry("MUX_CABLE", port, {"state": state_cfg_val,
                                                     "server_ipv4": ipv4_value, "server_ipv6": ipv6_value})
             port_status_dict[port] = 'OK'
-        if state_cfg_val == "auto":
-            # dont write anything to db, write OK to user
-            port_status_dict[port] = 'OK'
 
-    elif (state == "standby" and configdb_state == "auto") or (state == "unknown" and configdb_state == "auto"):
+    elif (state == "standby" and configdb_state == "auto") or (state == "unknown" and configdb_state == "auto") or (state == "standby" and configdb_state == "manual") or (state == "unknown" and configdb_state == "manual"):
```
Comment (jleveque) [inline config/muxcable.py:95]: Suggest simplifying:
```python
elif state in ["standby", "unknown"] and configdb_state in ["auto", "manual"]:
```

**Review by jleveque — CHANGES_REQUESTED** [review]

File: config/muxcable.py (hunk 2 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: config/muxcable.py (hunk 1 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: config/muxcable.py (hunk 3 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: config/muxcable.py (hunk 4 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed

File: config/muxcable.py (hunk 5 above)
Comment (vdahiya12) [reply config/muxcable.py]: fixed, thank you looks much more cleaner now.

Comment (jleveque) [reply config/muxcable.py]: You're welcome! Python offers some handy syntactic sugar :)

**Review by jleveque — APPROVED** [review]

File: config/muxcable.py [hunk 6]
```diff
@@ -73,43 +73,48 @@ def lookup_statedb_and_update_configdb(per_npu_statedb, config_db, port, state_c
     ipv6_value = get_value_for_key_in_config_tbl(config_db, port, "server_ipv6", "MUX_CABLE")
 
     state = get_value_for_key_in_dict(muxcable_statedb_dict, port, "state", "MUX_CABLE_TABLE")
-    if (state == "active" and configdb_state == "active") or (state == "standby" and configdb_state == "active") or (state == "unknown" and configdb_state == "active"):
-        if state_cfg_val == "active":
-            # status is already active, so right back error
+    if state in ["active", "standby", "unknown"] and configdb_state in ["active", "manual"]:
```
Comment (tahmed-dev) [inline config/muxcable.py:76]: can we simplify line 76 to line 180 as follows?

```
if state_cfg_val is not configdb_state:
    config_db.set_entry("MUX_CABLE", port, {"state": state_cfg_val})
    port_status_dict[port] = 'INPROGRESS' if state_cfg_val is 'active' and state is not 'active' else 'OK'
else:
     port_status_dict[port] = 'OK'
```

Comment (vdahiya12) [reply config/muxcable.py]: i think so, The logic is much more easier to follow. Thanks updated

File: config/muxcable.py [hunk 7]
```diff
@@ -74,7 +74,8 @@ def lookup_statedb_and_update_configdb(per_npu_statedb, config_db, port, state_c
 
     state = get_value_for_key_in_dict(muxcable_statedb_dict, port, "state", "MUX_CABLE_TABLE")
     if state_cfg_val is not configdb_state:
-        config_db.set_entry("MUX_CABLE", port, {"state": state_cfg_val})
+        config_db.set_entry("MUX_CABLE", port, {"state": "active",
```
Comment (tahmed-dev) [inline config/muxcable.py:77]: we can fix the lgtm be deleting the `ipv4_value` and `ip64_value`.

BTW, the logic could be simplified further as follows:

```
    if state_cfg_val is not configdb_state:
￼        config_db.set_entry("MUX_CABLE", port, {"state": state_cfg_val})		￼        
     port_status_dict[port] = 'INPROGRESS' if state_cfg_val is 'active' and state is not 'active' else 'OK'		￼        
```

Comment (tahmed-dev) [reply config/muxcable.py:77]: also, we are setting the entry to `active` instead of `state_cfg_val`.

File: config/muxcable.py [hunk 8]
```diff
@@ -73,43 +73,25 @@ def lookup_statedb_and_update_configdb(per_npu_statedb, config_db, port, state_c
...
 
-    elif (state == "standby" and configdb_state == "auto") or (state == "unknown" and configdb_state == "auto"):
-        if state_cfg_val == "active":
-            # make the state active
-            config_db.set_entry("MUX_CABLE", port, {"state": "active",
+    if str(state_cfg_val) == str(configdb_state):
+        port_status_dict[port] = 'OK'
+    else:
+        if str(state_cfg_val) == 'active' and str(state) != 'active':
+            config_db.set_entry("MUX_CABLE", port, {"state": state_cfg_val,
```
Comment (tahmed-dev) [inline config/muxcable.py:81]: the `config_db.set_entry` should be outside the if statement.

Comment (vdahiya12) [reply config/muxcable.py]: fixed

**Review by tahmed-dev — APPROVED** [review]

**Pull request merged**